File: src/boot/fe/extfmt.ml
(* The 'fmt' extension is modeled on the posix printf system.
 * 
 * A posix conversion ostensibly looks like this:
 * 
 * %[parameter][flags][width][.precision][length]type
 * 
 * Given the different numeric type bestiary we have, we omit the 'length'
 * parameter and support slightly different conversions for 'type':
 * 
 * %[parameter][flags][width][.precision]type
 * 
 * we also only support translating-to-rust a tiny subset of the possible
 * combinations at the moment.
 *)

exception Malformed of string
;;

type case =
    CASE_upper
  | CASE_lower
;;

type signedness =
    SIGNED
  | UNSIGNED
;;

type ty =
    TY_bool
  | TY_str
  | TY_char
  | TY_int of signedness
  | TY_bits
  | TY_hex of case
      (* FIXME: Support more later. *)
;;

type flag =
    FLAG_left_justify
  | FLAG_left_zero_pad
  | FLAG_left_space_pad
  | FLAG_plus_if_positive
  | FLAG_alternate
;;

type count =
    COUNT_is of int
  | COUNT_is_param of int
  | COUNT_is_next_param
  | COUNT_implied

type conv =
    { conv_parameter: int option;
      conv_flags: flag list;
      conv_width: count;
      conv_precision: count;
      conv_ty: ty }

type piece =
    PIECE_string of string
  | PIECE_conversion of conv


let rec peek_num (s:string) (i:int) (lim:int)
    : (int * int) option =
  if i >= lim
  then None
  else
    let c = s.[i] in
      if '0' <= c && c <= '9'
      then
        let n = (Char.code c) - (Char.code '0') in
          match peek_num s (i+1) lim with
              None -> Some (n, i+1)
            | Some (m, i) -> Some (n * 10 + m, i)
      else None
;;

let parse_parameter (s:string) (i:int) (lim:int)
    : (int option * int) =
  if i >= lim
  then (None, i)
  else
    match peek_num s i lim with
        None -> (None, i)
      | Some (n, j) ->
          if j < (String.length s) && s.[j] = '$'
          then (Some n, j+1)
          else (None, i)
;;

let rec parse_flags (s:string) (i:int) (lim:int)
    : (flag list * int) =
  if i >= lim
  then ([], i)
  else
    let cont flag =
      let (rest, j) = parse_flags s (i+1) lim in
        (flag :: rest, j)
    in
      match s.[i] with
          '-' -> cont FLAG_left_justify
        | '0' -> cont FLAG_left_zero_pad
        | ' ' -> cont FLAG_left_space_pad
        | '+' -> cont FLAG_plus_if_positive
        | '#' -> cont FLAG_alternate
        | _ -> ([], i)
;;

let parse_count (s:string) (i:int) (lim:int)
    : (count * int) =
  if i >= lim
  then (COUNT_implied, i)
  else
    if s.[i] = '*'
    then
      begin
        match parse_parameter s (i+1) lim with
            (None, j) -> (COUNT_is_next_param, j)
          | (Some n, j) -> (COUNT_is_param n, j)
      end
    else
      begin
        match peek_num s i lim with
            None -> (COUNT_implied, i)
          | Some (n, j) -> (COUNT_is n, j)
      end
;;

let parse_precision (s:string) (i:int) (lim:int)
    : (count * int) =
  if i >= lim
  then (COUNT_implied, i)
  else
    if s.[i] = '.'
    then parse_count s (i+1) lim
    else (COUNT_implied, i)
;;

let parse_type (s:string) (i:int) (lim:int)
    : (ty * int) =
  if i >= lim
  then raise (Malformed "missing type in conversion")
  else
    let t =
      match s.[i] with
          'b' -> TY_bool
        | 's' -> TY_str
        | 'c' -> TY_char
        | 'd' | 'i' -> TY_int SIGNED
        | 'u' -> TY_int UNSIGNED
        | 'x' -> TY_hex CASE_lower
        | 'X' -> TY_hex CASE_upper
        | 't' -> TY_bits
        | _ -> raise (Malformed "unknown type in conversion")
    in
      (t, i+1)
;;

let parse_conversion (s:string) (i:int) (lim:int)
    : (piece * int) =
  let (parameter, i) = parse_parameter s i lim in
  let (flags, i) = parse_flags s i lim in
  let (width, i) = parse_count s i lim in
  let (precision, i) = parse_precision s i lim in
  let (ty, i) = parse_type s i lim in
    (PIECE_conversion  { conv_parameter = parameter;
                         conv_flags = flags;
                         conv_width = width;
                         conv_precision = precision;
                         conv_ty = ty }, i)
;;

let parse_fmt_string (s:string) : piece array =
  let pieces = Queue.create () in
  let i = ref 0 in
  let lim = String.length s in
  let buf = Buffer.create 10 in
  let flush_buf _ =
    if (Buffer.length buf) <> 0
    then
      let piece =
        PIECE_string (Buffer.contents buf)
      in
        Queue.add piece pieces;
        Buffer.clear buf;
  in
    while (!i) < lim
    do
      if s.[!i] = '%'
      then
        begin
          incr i;
          if (!i) >= lim
          then raise (Malformed "unterminated conversion at end of string");
          if s.[!i] = '%'
          then
            begin
              Buffer.add_char buf '%';
              incr i;
            end
          else
            begin
              flush_buf();
              let (piece, j) = parse_conversion s (!i) lim in
                Queue.add piece pieces;
                i := j
            end
        end
      else
        begin
          Buffer.add_char buf s.[!i];
          incr i;
        end
    done;
    flush_buf ();
    Common.queue_to_arr pieces
;;


(*
 * Local Variables:
 * fill-column: 78;
 * indent-tabs-mode: nil
 * buffer-file-coding-system: utf-8-unix
 * compile-command: "make -k -C $RBUILD 2>&1 | sed -e 's/\\/x\\//x:\\//g'";
 * End:
 *)

File: src/lib/ExtFmt.rs
import option.none;
import option.some;

// Functions used by the fmt extension at compile time
mod CT {
    tag signedness {
        signed;
        unsigned;
    }

    tag caseness {
        case_upper;
        case_lower;
    }

    tag ty {
        ty_bool;
        ty_str;
        ty_char;
        ty_int(signedness);
        ty_bits;
        ty_hex(caseness);
        // FIXME: More types
    }

    tag flag {
        flag_left_justify;
        flag_left_zero_pad;
        flag_left_space_pad;
        flag_plus_if_positive;
        flag_alternate;
    }

    tag count {
        count_is(int);
        count_is_param(int);
        count_is_next_param;
        count_implied;
    }

    // A formatted conversion from an expression to a string
    type conv = rec(option.t[int] param,
                    vec[flag] flags,
                    count width,
                    count precision,
                    ty ty);

    // A fragment of the output sequence
    tag piece {
        piece_string(str);
        piece_conv(conv);
    }

    fn parse_fmt_string(str s) -> vec[piece] {
        let vec[piece] pieces = vec();
        auto lim = _str.byte_len(s);
        auto buf = "";

        fn flush_buf(str buf, &vec[piece] pieces) -> str {
            if (_str.byte_len(buf) > 0u) {
                auto piece = piece_string(buf);
                pieces += vec(piece);
            }
            ret "";
        }

        auto i = 0u;
        while (i < lim) {
            auto curr = _str.substr(s, i, 1u);
            if (_str.eq(curr, "%")) {
                i += 1u;
                if (i >= lim) {
                    log "unterminated conversion at end of string";
                    fail;
                }
                auto curr2 = _str.substr(s, i, 1u);
                if (_str.eq(curr2, "%")) {
                    i += 1u;
                } else {
                    buf = flush_buf(buf, pieces);
                    auto res = parse_conversion(s, i, lim);
                    pieces += vec(res._0);
                    i = res._1;
                }
            } else {
                buf += curr;
                i += 1u;
            }
        }
        buf = flush_buf(buf, pieces);
        ret pieces;
    }

    fn peek_num(str s, uint i, uint lim) -> option.t[tup(uint, uint)] {
        if (i >= lim) {
            ret none[tup(uint, uint)];
        }

        auto c = s.(i);
        if (!('0' as u8 <= c && c <= '9' as u8)) {
            ret option.none[tup(uint, uint)];
        }

        auto n = (c - ('0' as u8)) as uint;
        alt (peek_num(s, i + 1u, lim)) {
            case (none[tup(uint, uint)]) {
                ret some[tup(uint, uint)](tup(n, i + 1u));
            }
            case (some[tup(uint, uint)](?next)) {
                auto m = next._0;
                auto j = next._1;
                ret some[tup(uint, uint)](tup(n * 10u + m, j));
            }
        }

    }

    fn parse_conversion(str s, uint i, uint lim) -> tup(piece, uint) {
        auto parm = parse_parameter(s, i, lim);
        auto flags = parse_flags(s, parm._1, lim);
        auto width = parse_count(s, flags._1, lim);
        auto prec = parse_precision(s, width._1, lim);
        auto ty = parse_type(s, prec._1, lim);
        ret tup(piece_conv(rec(param = parm._0,
                               flags = flags._0,
                               width = width._0,
                               precision = prec._0,
                               ty = ty._0)),
                ty._1);
    }

    fn parse_parameter(str s, uint i, uint lim) -> tup(option.t[int], uint) {
        if (i >= lim) {
            ret tup(none[int], i);
        }

        auto num = peek_num(s, i, lim);
        alt (num) {
            case (none[tup(uint, uint)]) {
                ret tup(none[int], i);
            }
            case (some[tup(uint, uint)](?t)) {
                auto n = t._0;
                auto j = t._1;
                if (j < lim && s.(j) == '$' as u8) {
                    ret tup(some[int](n as int), j + 1u);
                }
                else {
                    ret tup(none[int], i);
                }
            }
        }
    }

    fn parse_flags(str s, uint i, uint lim) -> tup(vec[flag], uint) {
        let vec[flag] noflags = vec();

        if (i >= lim) {
            ret tup(noflags, i);
        }

        fn more_(flag f, str s, uint i, uint lim) -> tup(vec[flag], uint) {
            auto next = parse_flags(s, i + 1u, lim);
            auto rest = next._0;
            auto j = next._1;
            let vec[flag] curr = vec(f);
            ret tup(curr + rest, j);
        }

        auto more = bind more_(_, s, i, lim);

        auto f = s.(i);
        if (f == ('-' as u8)) {
            ret more(flag_left_justify);
        } else if (f == ('0' as u8)) {
            ret more(flag_left_zero_pad);
        } else if (f == (' ' as u8)) {
            ret more(flag_left_space_pad);
        } else if (f == ('+' as u8)) {
            ret more(flag_plus_if_positive);
        } else if (f == ('#' as u8)) {
            ret more(flag_alternate);
        } else {
            ret tup(noflags, i);
        }
    }

    fn parse_count(str s, uint i, uint lim) -> tup(count, uint) {
        if (i >= lim) {
            ret tup(count_implied, i);
        }

        if (s.(i) == ('*' as u8)) {
            auto param = parse_parameter(s, i + 1u, lim);
            auto j = param._1;
            alt (param._0) {
                case (none[int]) {
                    ret tup(count_is_next_param, j);
                }
                case (some[int](?n)) {
                    ret tup(count_is_param(n), j);
                }
            }
        } else {
            auto num = peek_num(s, i, lim);
            alt (num) {
                case (none[tup(uint, uint)]) {
                    ret tup(count_implied, i);
                }
                case (some[tup(uint, uint)](?num)) {
                    ret tup(count_is(num._0 as int), num._1);
                }
            }
        }
    }

    fn parse_precision(str s, uint i, uint lim) -> tup(count, uint) {
        if (i >= lim) {
            ret tup(count_implied, i);
        }

        if (s.(i) == '.' as u8) {
            ret parse_count(s, i + 1u, lim);
        } else {
            ret tup(count_implied, i);
        }
    }

    fn parse_type(str s, uint i, uint lim) -> tup(ty, uint) {
        if (i >= lim) {
            log "missing type in conversion";
            fail;
        }

        auto t;
        auto tstr = _str.substr(s, i, 1u);
        if (_str.eq(tstr, "b")) {
            t = ty_bool;
        } else if (_str.eq(tstr, "s")) {
            t = ty_str;
        } else if (_str.eq(tstr, "c")) {
            t = ty_char;
        } else if (_str.eq(tstr, "d")
                   || _str.eq(tstr, "i")) {
            // TODO: Do we really want two signed types here?
            // How important is it to be printf compatible?
            t = ty_int(signed);
        } else if (_str.eq(tstr, "u")) {
            t = ty_int(unsigned);
        } else if (_str.eq(tstr, "x")) {
            t = ty_hex(case_lower);
        } else if (_str.eq(tstr, "X")) {
            t = ty_hex(case_upper);
        } else if (_str.eq(tstr, "t")) {
            t = ty_bits;
        } else {
            log "unknown type in conversion";
            fail;
        }

        ret tup(t, i + 1u);
    }
}

// Functions used by the fmt extension at runtime
mod RT {
    fn conv_int(int i) -> str {
        ret _int.to_str(i, 10u);
    }

    fn conv_uint(uint u) -> str {
        ret _uint.to_str(u, 10u);
    }

    fn conv_bool(bool b) -> str {
        if (b) {
            ret "true";
        } else {
            ret "false";
        }
    }

    fn conv_char(char c) -> str {
        ret _str.from_char(c);
    }
}

// Local Variables:
// mode: rust;
// fill-column: 78;
// indent-tabs-mode: nil
// c-basic-offset: 4
// buffer-file-coding-system: utf-8-unix
// compile-command: "make -k -C .. 2>&1 | sed -e 's/\\/x\\//x:\\//g'";
// End:
